### ISSUE TITLE
isal_zstate.next_in pointer at wrong position after inflating with isal_inflate using non-gzip crc_flags

### ISSUE THREAD

**Issue by rhpvorderman** [intel/isa-l#157]
I came across this when working on the Python api. I decided to mimick the python zlib api using ISA-L, since python's `gzip` module uses the `zlib` module to implement its functions.

The zlib api for python can be found [here](https://docs.python.org/3.10/library/zlib.html). It features a `Compress` and `Decompress` object, which are basically wrappers for a `z_stream` (or `isal_zstream` in this case). These objects can be fed data and will update their internal z_stream object. These objects are used for streaming implementations of zlib/gzip compatible compression/decompression.

I had some struggles implementing the `Decompress` object correctly. It features `unconsumed_tail` and `unused_data` members. These members are used to report back the data that was **not** decompressed on the last decompress call. I had problems with `unused_data`. This member is populated when an end of a compressed stream is reached. Since gzip allows concatenating various members together, the `unused_data` will contain the header of the next gzip member.

In theory this is quite easy. As soon as `inflate_state.block_state` is `ISAL_BLOCK_FINISH`, compression is stopped and `inflate_state.next_in` should point at the first character after the compressed block.

See the following test. The data is compressed and some nonsense data `abcdefghijklmnopqrstuvwxyz` is appended to the compressed block. After decompressing we should get the data back and also the full nonsense data.

```python
def test_unused_data(wbits):
    unused_data = b"abcdefghijklmnopqrstuvwxyz"
    compressor = zlib.compressobj(wbits=wbits)
    data = b"A meaningful sentence starts with a capital and ends with a."
    compressed = compressor.compress(data) + compressor.flush()
    decompressor = isal_zlib.decompressobj(wbits=wbits)
    result = decompressor.decompress(compressed + unused_data)
    assert result == data
    assert decompressor.unused_data == unused_data
```

This does not work with all crc_flags for isa-l. When the crc_flag == ISAL_GZIP or ISAL_GZIP_NO_HDR_VER. It works.
When the `crc_flag == ISAL_ZLIB` the data comes back correctly but the unused_data comes back `defghijklmnopqrstuvwxyz`. When the `crc_flag == ISAL_DEFLATE` the data also comes back correctly but the unused_data comes back as `hijklmnopqrstuvwxyz`. 

This is due to the pointer position being slightly off. For `ISAL_GZIP` `next_in` points precisely at the first character after the compressed block, but for the non-gzip crc_flags this is not the case. Meddling with the offsets is no good. Setting the next_in pointer back 7 characters for ISAL_DEFLATE for instance will give wrong results when there is no unused data (and therefore the last 7 characters of the last deflate block will be reported).

Is this a fixable error? Or is there some oversight on my end with regards to the next_in pointer on an inflate_state?

**Comment by gbtucker** [comment]
It is possible for decompression to read past the eob and trailer but not past the buffer size.  This data is buffered internally and can be used in the next call unless you are using isal_inflate_stateless().  The number of bytes should be (state.read_in_length / 8).

**Comment by rhpvorderman** [comment]
I used that. They should be in the "read_in" 64-bit integer right? I tried to get results from that, but was not succesful in my implementation. But come to think of it. I did notice that the maximum amount of overshoot I saw was 7. So indeed everything should be recoverable from the read_in. I just to make sure the results are correctly interpreted.
It will require some fiddling to get this working. But thanks for the pointers, I at least know that that way is on the right track.

Having said that. I will be on hiatus until the middle of November. So the python bindings will have to wait until after that. Thanks for your help so far! I hope I will be able to release some python bindings before the end of the year.

**Comment by rhpvorderman** [comment]
I think this issue can be closed:

 Only gzip supports multiple members in the same archive, each with its own header. Therefore it is important that inflate_state.next_in is correct after decompressing a block. Which it is.

In  this particular use case which are the python bindings, I mimicked gzip.py's behaviour of reading in the header in python. Then decompressing the deflate block using isa-l and then reading the trailer again in python. This has the advantage of being able to verify that the data has been decompressed and stored correctly in python. It is a bit of an odd use case and it has been solved using the `read_in` buffer together with `read_in_length`. Python-isal versions 0.7.0 and higher have this fix implemented and provide almost one-to-one replacements of the python zlib and gzip modules.

I don't think a fix is needed for other people using ISA-L as using the "ISAL_GZIP" crc flag is perfectly viable for most use cases.
